homeassistant/components/abode/__init__.py

```diff
@@ -96,83 +94,84 @@
-        """Check if a switch device is configured as a light."""
-        import abodepy.helpers.constants as CONST
+    conf = config[DOMAIN]
 
-        return device.generic_type == CONST.TYPE_LIGHT or (
-            device.generic_type == CONST.TYPE_SWITCH and device.device_id in self.lights
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": SOURCE_IMPORT}, data=deepcopy(conf)
         )
+    )
 
+    return True
 
-def setup(hass, config):
-    """Set up Abode component."""
-    from abodepy.exceptions import AbodeException
 
-    conf = config[DOMAIN]
-    username = conf.get(CONF_USERNAME)
-    password = conf.get(CONF_PASSWORD)
-    name = conf.get(CONF_NAME)
-    polling = conf.get(CONF_POLLING)
-    exclude = conf.get(CONF_EXCLUDE)
-    lights = conf.get(CONF_LIGHTS)
+async def async_setup_entry(hass, config_entry):
+    """Set up Abode component via config entry (Integrations UI)."""
+    username = config_entry.data.get(CONF_USERNAME)
+    password = config_entry.data.get(CONF_PASSWORD)
+    polling = config_entry.data.get(CONF_POLLING)
 
     try:
         cache = hass.config.path(DEFAULT_CACHEDB)
-        hass.data[DOMAIN] = AbodeSystem(
-            username, password, cache, name, polling, exclude, lights
+        abode = Abode(
+            username,
+            password,
+            auto_login=True,
+            get_devices=True,
+            get_automations=True,
+            cache_path=cache,
+        )
+        hass.data[DOMAIN] = await hass.async_add_executor_job(
+            AbodeSystem, abode, username, password, cache, polling
         )
     except (AbodeException, ConnectTimeout, HTTPError) as ex:
         _LOGGER.error("Unable to connect to Abode: %s", str(ex))
+        return False
 
-        hass.components.persistent_notification.create(
-            "Error: {}<br />"
-            "You will need to restart hass after fixing."
-            "".format(ex),
-            title=NOTIFICATION_TITLE,
-            notification_id=NOTIFICATION_ID,
+    for platform in ABODE_PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(config_entry, platform)
         )
-        return False
 
-    setup_hass_services(hass)
-    setup_hass_events(hass)
-    setup_abode_events(hass)
+    await setup_hass_events(hass)
+    await hass.async_add_executor_job(setup_hass_services, hass)
+    await hass.async_add_executor_job(setup_abode_events, hass)
+
+    return True
+
+
+async def async_unload_entry(hass, config_entry):
+    """Unload a config entry."""
+    hass.services.async_remove(DOMAIN, SERVICE_SETTINGS)
+    hass.services.async_remove(DOMAIN, SERVICE_CAPTURE_IMAGE)
+    hass.services.async_remove(DOMAIN, SERVICE_TRIGGER)
 
     for platform in ABODE_PLATFORMS:
-        discovery.load_platform(hass, platform, DOMAIN, {}, config)
+        await hass.config_entries.async_forward_entry_unload(config_entry, platform)
```

Comment:
We can use asyncio.wait or asyncio.gather to await all coroutines concurrently.

Reply:
This is new to me (much like the rest of Python) but this is what I gather after doing a bit of research and just wanted to make sure I have this correct:
```
from asyncio import gather
````
Then:

```
    for platform in ABODE_PLATFORMS:
        await gather(
            hass.config_entries.async_forward_entry_unload(config_entry, platform)
        )
```

Reply:
We need to collect the coroutine objects in a list first. Then use gather on the items in the list.

Reply:
Ah makes sense - I had a feeling that was the case reading some of the docs. Something like this:

```
    tasks = []

    for platform in ABODE_PLATFORMS:
        tasks.append(
            hass.config_entries.async_forward_entry_unload(config_entry, platform)
        )

    await gather(*tasks)
```